File: modules/system.py
"""
System Module
Controls macOS system functions via AppleScript
"""

import asyncio
import subprocess
import re

class SystemModule:
    def __init__(self, core):
        self.core = core
        self.name = 'system'
        self.description = 'System control and information'
        
    async def execute(self, message, decision):
        """Execute system command"""
        message_lower = message.lower()
        entities = decision.get('entities', {})
        
        try:
            if 'open' in message_lower:
                app = entities.get('app', self._extract_app_name(message))
                if app:
                    await self._run_applescript(f'tell application "{app}" to activate')
                    return {'text': f'Opening {app}', 'type': 'success'}
                    
            elif 'close' in message_lower:
                app = entities.get('app', self._extract_app_name(message))
                if app:
                    await self._run_applescript(f'tell application "{app}" to quit')
                    return {'text': f'Closing {app}', 'type': 'success'}
                    
            elif 'volume' in message_lower:
                numbers = entities.get('numbers', [])
                if numbers:
                    volume = numbers[0]
                    await self._run_applescript(f'set volume output volume {volume}')
                    return {'text': f'Volume set to {volume}%', 'type': 'success'}
                    
            elif 'screenshot' in message_lower:
                await self._run_applescript('do shell script "screencapture -c"')
                return {'text': 'Screenshot taken to clipboard', 'type': 'success'}
                
            elif 'lock' in message_lower:
                await self._run_applescript('do shell script "/System/Library/CoreServices/Menu\\\\ Extras/User.menu/Contents/Resources/CGSession -suspend"')
                return {'text': 'Locking screen', 'type': 'success'}
                
            return {'text': 'Could not understand system command', 'type': 'error'}
            
        except Exception as e:
            return {'text': f'Error: {str(e)}', 'type': 'error'}
            
    async def _run_applescript(self, script):
        """Run AppleScript command"""
        escaped_script = script.replace("'", "'\\''")
        process = await asyncio.create_subprocess_shell(
            f"osascript -e '{escaped_script}'",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
        stdout, stderr = await process.communicate()
        
        if stderr:
            raise Exception(stderr.decode())
            
        return stdout.decode().strip()
        
    def _extract_app_name(self, message):
        """Extract application name from message"""
        apps = ['Safari', 'Chrome', 'Firefox', 'Spotify', 'Music', 'Notes', 'Mail', 'Calendar', 'Finder']
        message_lower = message.lower()
        
        for app in apps:
            if app.lower() in message_lower:
                return app
                
        return None

File: modules/system.js
class SystemModule {
  constructor(core) {
    this.core = core;
    this.name = 'system';
    this.description = 'System control and information';
  }

  canHandle(message) {
    const lower = message.toLowerCase();
    return lower.includes('open') || 
           lower.includes('close') || 
           lower.includes('volume') ||
           lower.includes('brightness') ||
           lower.includes('screenshot') ||
           lower.includes('lock') ||
           lower.includes('sleep');
  }

  async execute(message) {
    const lower = message.toLowerCase();

    try {
      if (lower.includes('open')) {
        const app = this.extractAppName(message);
        if (app) {
          await this.core.executeAppleScript(`tell application "${app}" to activate`);
          return { text: `Opening ${app}`, type: 'success' };
        }
      }

      if (lower.includes('close')) {
        const app = this.extractAppName(message);
        if (app) {
          await this.core.executeAppleScript(`tell application "${app}" to quit`);
          return { text: `Closing ${app}`, type: 'success' };
        }
      }

      if (lower.includes('volume')) {
        const volume = this.extractNumber(message);
        if (volume !== null) {
          await this.core.executeAppleScript(`set volume output volume ${volume}`);
          return { text: `Volume set to ${volume}%`, type: 'success' };
        }
      }

      if (lower.includes('screenshot')) {
        await this.core.executeAppleScript(`do shell script "screencapture -c"`);
        return { text: 'Screenshot taken to clipboard', type: 'success' };
      }

      if (lower.includes('lock')) {
        await this.core.executeAppleScript(`do shell script "/System/Library/CoreServices/Menu\\\\ Extras/User.menu/Contents/Resources/CGSession -suspend"`);
        return { text: 'Locking screen', type: 'success' };
      }

      return { text: 'Could not understand system command', type: 'error' };
    } catch (error) {
      return { text: `Error: ${error.message}`, type: 'error' };
    }
  }

  extractAppName(message) {
    const apps = ['Safari', 'Chrome', 'Firefox', 'Spotify', 'Music', 'Notes', 'Mail', 'Calendar', 'Messages', 'Finder', 'Terminal'];
    const lower = message.toLowerCase();
    
    for (const app of apps) {
      if (lower.includes(app.toLowerCase())) {
        return app;
      }
    }
    
    return null;
  }

  extractNumber(message) {
    const match = message.match(/\d+/);
    return match ? parseInt(match[0]) : null;
  }
}

module.exports = SystemModule;
